feat: add blog post for templates v0.8 release (#227)
arcus-azure/arcus: community/blog/2022-06 .NET 6 finally comes to Arcus Templates in v0.8 release/README.md

# .NET 6 finally comes to Arcus Templates in v0.8 release, together with new templates and features

> The v0.8 release of Arcus Templates ends the journey of adding .NET support to Arcus. This release also contains tons of new features and useful changes. This run-down post will guide you through the biggest ones.

Reviewer: When I read this first sentence, it's like this is the last release where we support .NET.
I would rewrite this to something more like 

> With the v0.8 release of Arcus Templates, we now fully support .NET 6.0 in Arcus.

Or something better :)

Author: > When I read this first sentence, it's like this is the last release where we support .NET.

🤔 that's a different take. Seems like 'X comes to Y' that it is finally there, not leaving 😄 .

Author: The proposed title doesn't include the other stuff, and since .NET 6 is only < 1/3 of the post, it doesn't seem right to only include that in the title.
`Arcus Templates v0.8 finally brings .NET 6 support to Arcus, together with new templates and features`. Something like this?

Reviewer: Yes, that is ok.

## New Azure Functions Service Bus message handling project templates

> The core functionality of the Arcus Messaging library; is message handling. Describing the business logic in reusable message handlers where an incoming message is routed via Arcus functionality to the correct handler. Previous releases of Arcus Messaging made sure that the functionality that receives a message is separated from the functionality that routes the message. The reason for this is so we can reuse the message routing in different scenarios. This is one of those scenarios.

Reviewer 2: ```suggestion
The core functionality of the Arcus Messaging library is message handling. Describing the business logic in reusable message handlers where an incoming message is routed via Arcus functionality to the correct handler. Previous releases of Arcus Messaging made sure that the functionality that receives a message is separated from the functionality that routes the message. The reason for this is so we can reuse the message routing in different scenarios. This is one of those scenarios.
```

> The core functionality of the Arcus Messaging library; is message handling. Describing the business logic in reusable message handlers where an incoming message is routed via Arcus functionality to the correct handler. Previous releases of Arcus Messaging made sure that the functionality that receives a message is separated from the functionality that routes the message. The reason for this is so we can reuse the message routing in different scenarios. This is one of those scenarios.

Reviewer: ```suggestion
The core functionality of the Arcus Messaging library, is message handling. Describing the business logic in reusable message handlers where an incoming message is routed via Arcus functionality to the correct handler. Previous releases of Arcus Messaging made sure that the functionality that receives a message is separated from the functionality that routes the message. The reason for this is so we can reuse the message routing in different scenarios. This is one of those scenarios.
```

Author: Already removed comma. Grammerly suggested a colon, though 😉 .

> Azure Functions differ from the usual message pump setup because the message is already being received by the trigger function; whether it is a message on an Azure Service Bus queue or a topic subscription. Since the trigger function already takes over the work of the message pump, we only need to route messages.

Reviewer: ```suggestion
Azure Functions differ from the usual message pump setup because the message is already being received by the function's trigger; whether it is a message on an Azure Service Bus queue or a topic subscription. Since the trigger already takes over the work of the message pump, we only need to route messages.
```

> Easy extensions on the Azure Functions types allow you to register your message handlers the same way as with a message pump setup. 

Reviewer: ```suggestion
Easy extensions on the `IFunctionsHostBuilder` allow you to register your message handlers the same way as with a message pump setup. 
```

Author: Yes, was thinking to include this, although, it is not the only type that is being extended (or would be extended), so not sure to mention it?

Reviewer: But it is maybe the main type ? Maybe this then ?

Author: Yeah, currently it's still only this type, so I applied your suggestion. 😉 .

using Arcus.Messaging.Abstractions.ServiceBus;
using Arcus.Messaging.Abstractions.ServiceBus.MessageHandling;
using Azure.Messaging.ServiceBus;
using Microsoft.Azure.ServiceBus;
using Microsoft.Azure.WebJobs;
using Microsoft.Extensions.Logging;

public class OrderFunction
{
    private readonly IAzureServiceBusMessageRouter _messageRouter;
    private readonly string _jobId;

    public OrderFunction(IAzureServiceBusMessageRouter messageRouter)
    {
        _messageRouter = messageRouter;
        _jobId = Guid.NewGuid().ToString();
    }

    [FunctionName("order-processing")]
    public async Task Run(
        [ServiceBusTrigger("orders", Connection = "ServiceBusConnectionString")] ServiceBusReceivedMessage message,
        ILogger log,
        CancellationToken cancellationToken)
    {
        log.LogInformation($"C# ServiceBus queue trigger function processed message: {message.MessageId}");

        AzureServiceBusMessageContext context = message.GetMessageContext(_jobId);
        MessageCorrelationInfo correlationInfo = message.GetCorrelationInfo();

        await _messageRouter.RouteMessageAsync(message, context, correlationInfo, cancellationToken);

>     }

Reviewer: So, this is how we should use the Arcus.Messaging functionality in an Azure Function.
Do we have this information in our documentation as well ?

Author: The messaging documentation has this explained on a separated page: https://messaging.arcus-azure.net/Features/message-pumps/service-bus-azure-functions

Author: Resolving for now.

}
```

The code described here is an exact implementation of what is available in the new Arcus Azure Functions Service Bus project templates. We have one for both [Azure Service Bus queue](https://templates.arcus-azure.net/features/azurefunctions-servicebus-queue-template) and [Azure Service Bus topic subscription](https://templates.arcus-azure.net/features/azurefunctions-servicebus-topic-template).

## Remove exception details from health report in Web API project template

> The Swagger OpenAPI documentation is by default available in the Web API project template (but can be excluded if needed). Adding Swagger to your API also means that the API models used during request/response will be externalized in the Swagger endpoint JSON. It was by inspecting this Swagger result that we saw a problem with the default application health endpoint that is available in the Web API project template. This endpoint used Microsoft's `HealthReport` model which can include exception details when the application is unhealthy. Generating the Swagger document caused failures because the exception details could contain possible internal type information which resulted in an enormous amount of types that needed to be included in the document. Also from a security perspective, this is not a good approach because it leaks information about the internal system. It could even leak application types that aren't used in the API.

Reviewer 2: ```suggestion
The OpenAPI documentation is by default available in the Web API project template (but can be excluded if needed). Adding the OpenAPI documentation to your API also means that the API models used during request/response will be externalized in the OpenAPI endpoint JSON. It was by inspecting this OpenAPI documentation that we saw a problem with the default application health endpoint that is available in the Web API project template. This endpoint used Microsoft's `HealthReport` model which can include exception details when the application is unhealthy. Generating the OpenAPI document caused failures because the exception details could contain possible internal type information which resulted in an enormous amount of types that needed to be included in the document. Also from a security perspective, this is not a good approach because it leaks information about the internal system. It could even leak application types that aren't used in the API.
```
I changed Swagger to OpenAPI as Swagger is more the tool than the actual API documentation, see also https://swagger.io/blog/api-strategy/difference-between-swagger-and-openapi/.

Author: Aha, ok, thx for this!

The health endpoint available in the Web API project template uses this new API model as a way to expose the application's health. For more information on the Web API project template, see [our official documentation](https://templates.arcus-azure.net/features/web-api-template).

## Hide JSON serialization behind maintainable extensions
The Arcus templates are especially useful for speeding up development and removing the need to write boilerplate code. Message handling, secret management, we have all sorts of ways to abstract a problem into business components. Sometimes, though, it is not such a big piece of boilerplate functionality. The Web API project template used to have some lines of code to restrict and configure the JSON serialization. It was not a lot but it was enough to make the application startup code dirty. Not only dirty; but less maintainable. The code was removing output formatters, changing how the JSON serialization handles enumerations, etc. If we would decide to change this strategy, we couldn't provide support to consumers who already ran the template in their project.

> A previous Web API release made sure that our JSON serialization strategy is accessible through the MVC options, which makes it not only cleaner but also more maintainable.

Reviewer 2: I don't completely understand this line, does this mean this was already solved in a previous release? Why would we mention it in the v0.8 release then?

Author: Well, it was the Web API release that made these extensions available, so it was solved in that release, but wasn't used until now.

Reviewer 2: Ah I understand, thanks for the clarification

```csharp
WebApplicationBuilder builder = ...

builder.Services.AddControllers(options =>
{
    options.OnlyAllowJsonFormatting();
    options.ConfigureJsonFormatting(json =>
    {
        json.DefaultIgnoreCondition = JsonIgnoreCondition.WhenWritingNull;
        json.Converters.Add(new JsonStringEnumConverter());
    });
});
```

> The code sample is the actual result of how the Web API project template now handles JSON serialization. As you can see, it is a much cleaner and safer approach.

Author: Yeah, was thinking that too, or a diff of some kind. It would however make it a bit complex.
I'll check.

> The code sample is the actual result of how the Web API project template now handles JSON serialization. As you can see, it is a much cleaner and safer approach.

Author: Added in https://github.com/arcus-azure/arcus/pull/227/commits/0d0c91e4892f62d4bb6271c506ca0fe3a9085bef
